### DISCUSSION TITLE
build mitmproxy binaries for arm64 linux

### DISCUSSION THREAD

**Discussion by pkill37** [mitmproxy/mitmproxy#5632] [Q&A] [answered]
I want to build one of these precompiled binaries found in https://mitmproxy.org/downloads/#branches/main/, such as mitmproxy-main-linux.tar.gz, but for arm64 Linux

@mhils pointed me to https://github.com/mitmproxy/mitmproxy/blob/main/release/build.py#L120 but it's still not very clear to me how to proceed

With more pointers perhaps I could contribute the code back 😄

**Comment by mhils** [comment]
1. We already provide arm64 Docker images which may be of interest to you: https://hub.docker.com/r/mitmproxy/mitmproxy/tags
2. Our binaries are automatically built by our CI systems. The corresponding job definition is here: https://github.com/mitmproxy/mitmproxy/blob/a1631154d722ea4c82dcb72e0c84ad38c6991de5/.github/workflows/main.yml#L106-L144

I think the main blocker right now is that GitHub Actions does not provide arm64 runners (yet), which makes building tricky. I haven't looked closer though.

**Comment by pkill37** [comment]
Thanks! I need a flat binary solution so I went ahead and tried to build it myself directly on Ubuntu 20.04 on an arm64 machine like this

```
apt install -y python3.9 python3.9-dev python3-pip python3-venv git
git clone https://github.com/mitmproxy/mitmproxy.git /tmp/mitmproxy
cd /tmp/mitmproxy
python3.9 -m pip install -e .[dev] && python3.9 -u release/build.py standalone-binaries
```

I take the release from `./release/dist/mitmproxy-dev-linux.tar.gz` and deploy those executables and run them on the same arm64 machine but on Debian 11

```
$ mitmweb
Traceback (most recent call last):
  File "mitmweb", line 2, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "mitmproxy/tools/main.py", line 10, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "mitmproxy/master.py", line 6, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "mitmproxy/addonmanager.py", line 14, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "mitmproxy/flow.py", line 6, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "mitmproxy/connection.py", line 8, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "mitmproxy/certs.py", line 17, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "OpenSSL/__init__.py", line 8, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "OpenSSL/crypto.py", line 11, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "OpenSSL/_util.py", line 5, in <module>
  File "PyInstaller/loader/pyimod02_importers.py", line 499, in exec_module
  File "cryptography/hazmat/bindings/openssl/binding.py", line 14, in <module>
ImportError: libffi-2a6f5b63.so.8.1.0: cannot open shared object file: No such file or directory
[5611] Failed to execute script 'mitmweb' due to unhandled ### exception!
```

Doesn't it statically link libffi? Why am I having this problem? 🤔 

```
$ ldd $(which mitmweb)
        linux-vdso.so.1 (0x0000e364c79c5000)
        libdl.so.2 => /lib/aarch64-linux-gnu/libdl.so.2 (0x0000e364c7981000)
        libz.so.1 => /lib/aarch64-linux-gnu/libz.so.1 (0x0000e364c7957000)
        libpthread.so.0 => /lib/aarch64-linux-gnu/libpthread.so.0 (0x0000e364c7926000)
        libc.so.6 => /lib/aarch64-linux-gnu/libc.so.6 (0x0000e364c77b1000)
        /lib/ld-linux-aarch64.so.1 (0x0000e364c7995000)
```

**Comment by pkill37** [comment]
I tried building on Debian 11 with the same result

**Comment by pkill37** [comment]
Got it! `strace` pointed out to me that this flat binary magic is doing everything in a `/tmp` directory which coincidentally I did not have permissions to read from in my particular setup

```
$ strace mitmweb
...
unlinkat(AT_FDCWD, "/tmp/_MEIb2eRQU/lib-dynload/readline.cpython-39-aarch64-linux-gnu.so", 0) = 0
newfstatat(AT_FDCWD, "/tmp/_MEIb2eRQU/lib-dynload/resource.cpython-39-aarch64-linux-gnu.so", {st_mode=S_IFREG|0700, st_size=15032, ...}, AT_SYMLINK_NOFOLLOW) = 0
unlinkat(AT_FDCWD, "/tmp/_MEIb2eRQU/lib-dynload/resource.cpython-39-aarch64-linux-gnu.so", 0) = 0
newfstatat(AT_FDCWD, "/tmp/_MEIb2eRQU/lib-dynload/_codecs_kr.cpython-39-aarch64-linux-gnu.so", {st_mode=S_IFREG|0700, st_size=137440, ...}, AT_SYMLINK_NOFOLLOW) = 0
unlinkat(AT_FDCWD, "/tmp/_MEIb2eRQU/lib-dynload/_codecs_kr.cpython-39-aarch64-linux-gnu.so", 0) = 0
getdents64(4, 0x3dec5f0 /* 0 entries */, 32768) = 0
close(4)                                = 0
unlinkat(AT_FDCWD, "/tmp/_MEIb2eRQU/lib-dynload", AT_REMOVEDIR) = 0
newfstatat(AT_FDCWD, "/tmp/_MEIb2eRQU/libffi.so.7", {st_mode=S_IFREG|0700, st_size=35224, ...}, AT_SYMLINK_NOFOLLOW) = 0
unlinkat(AT_FDCWD, "/tmp/_MEIb2eRQU/libffi.so.7", 0) = 0
getdents64(3, 0x3de45b0 /* 0 entries */, 32768) = 0
brk(0x3e05000)                          = 0x3e05000
close(3)                                = 0
unlinkat(AT_FDCWD, "/tmp/_MEIb2eRQU", AT_REMOVEDIR) = 0
brk(0x3e00000)                          = 0x3e00000
exit_group(1)                           = ?
+++ exited with 1 +++
```

**Comment by JohnBernardsson** [comment]
Hello! Not sure if this is the best place to ask, but... I am a python newbie, and I wonder if is there some documentation on how to build the binaries locally?

I did some modifications on the map local addon, to support regex capturing groups & replacement, basically to use mitmproxy as a replacement of Fiddler. I have been using it flawlessly for months now, and some folks in the company are interested in testing it too. But I think it would be nice to send them some binaries, instead of instructing them to clone the github branch, do the dev install, and use it in dev mode.

I am on a Windows machine, and following what @darksonic37 did, I just tried to do: `python -u release/build.py standalone-binaries`. But that fails miserabily with:

```
Traceback (most recent call last):
  File "C:\Users\juanb\Dev\github\johnbernardsson-mitmproxy\release\build.py", line 17, in <module>
    import click
ModuleNotFoundError: No module named 'click'
```

That seems to me like some pretty basic thing I am obviating 😅, so I am totally lost. Any directions you guys can give me?

For context: @mhils I mentioned this to you in slack, but you mentioned some issue with... backtracking? (I don't remember exactly, and the conversation is too old to check in the slack channel). You also told me probably it could be done with some combination of map-remote and map-local, but I spend some time one afternoon trying it, and I wasn't able to 😞.
So as my changes are working very well for my use case, I just continued with them and didn't look back... 😅

**Reply by mhils** [reply]
First off, feel free to just open a PR with your changes and I'll take a look. Slack is bad for this because it's not indexed by search engines (and I have the memory of a fruit fly). Second, you need to install the dev dependencies `pip install -e .[dev]`, which include click.

**Reply by JohnBernardsson** [reply]
Thanks a lot @mhils! For some reason I thought that already was done with the `venv\Scripts\pip install -e .[dev]` but of course, that was only for the virtual environment 🤦 

I have been able to do the release build succesfully!

Now that I have the changes in a github branch, and that today I was able to run the tests locally, I will try to clean the changes (some tests are failing), and also see what is happening with the binary, as it is giving me an error that doesn't happen with the venv dev version. When I have that done, I can create a PR (with great shame, because I am sure I am doing terrible things there... this is my first python development 😅).

Thanks for your help and your quick response 😃 !

**Comment by iMonZ** [comment]
Hey any news here?

**Comment by mhils** [comment, marked as answer]
https://mitmproxy.org/downloads/ has arm64 binaries nowadays.
